rfcs/0074-tvmscript-unified-printer.md

```diff
@@ -0,0 +1,487 @@
+Relax and any future vendor-specific IR are all treated equally as dialects and
+could be printed together.
+
+# Motivation
+[motivation]: #motivation
+
+TVMScript, as a roundtrippable python-based text format, is
+the central piece of TVM performance productivity. As the frontend of TVM, it
+enables end users to directly construct the TVM IR, either TIR or Relax, in a
+pragmatic approach. From Relax to MetaSchedule and TIR, TVMScript enables
+inspectability and reproducibility at any level of compilation and
+optimization. Furthermore, based on TVMScript, developers are empowered to
+intercept, manipulate and customize the compiler behavior in a principled way.
+
+While TVMScript is gaining traction and buy-in from the open source community,
+the TVMScript printer suffers from multiple profound design issues:
+Not supporting IR fragment printing requires users to jump in-between
+TVMScript syntax and TIRText syntax; The lack of modularity leads to practical
+inability to scale up to and maintain multiple IRs without engineering
+conflicts; Enhancing co-existence of multi-level IRs often leads to
+re-engineering of existing features.
+
+**Goal.** This RFC introduces Tvmscript UNIfied Printer (TUNIP), a systematic
+redesign to
+address those engineering, usability and scalability issues above. The goal of
+this re-design includes:
+
+**Goal 1 [Unified Representation].** Become the unified roundtrippable
+representation of TIR and Relax, allowing systematic mixing of IRs or IR
+fragments (Relax + TIR) in the same IRModule in the target language (for
+example, python, C++). 
+
+The current TVMScript printer only supports TIR. An experimental printer is
+being developed at
+[https://github.com/tlc-pack/relax/blob/relax/src/printer/relax_script_printer.cc](https://github.com/tlc-pack/relax/blob/relax/src/printer/relax_script_printer.cc)
+for Relax, which needs to handle cases where TIR is embedded in Relax (for
+instance, https://github.com/tlc-pack/relax/pull/149). By having a unified
+approach for printing IR tree to TVMScript, we will be able to implement a
+fully-fledged printer for Relax and TIR with minimal effort.
+
+**Goal 2 [Third-Party IRs in Multi-Stage Compilation].** Modularize and
+infrastructuralize the printer to support more future IRs or third-party IRs at
+any level with maintainability, for example, IRs at lower-level than TIR, or
+Relax VM executable.
+
+The current TVMScript printer is tightly coupled with TIR by being a subclass
+of TIR-specific functors. This design isn’t scalable when we want to support
+more IRs
+
+```cpp
+class TVMScriptPrinter : public tir::StmtFunctor<Doc(const tir::Stmt&)>,
+                         public tir::ExprFunctor<Doc(const PrimExpr&, ExprPrecedence*)>,
+                         public TypeFunctor<Doc(const Type&)>  
+{...}
+```
+
+More importantly, it’s impossible for the current approach to support
+third-party IR bteing registered in a dynamic library.
```

Comment:
A second thought: example for the existing work is not necessary. Just put a link here for people who are interested to see, while we focus on our new design

Reply:
Updated